### PR TITLE
Feature/create intake subsystem

### LINKED ISSUE #21: Create Intake subsystem
Create the intake subsystem.

Should be driven by 1 motor. Method needed for setting speed.
When initialized (and always) motor should be in Brake mode.

Also has 1 double solenoid. Methods needed for `neutral`, `extend`, `retract`, and `toggle`.

Togggle should go:
extend -> retract
retract -> extend
neutral -> retract

When retracted, calls to set the speed (to a speed other than 0) should throw an IllegalStateException.
When the solenoid is set to retracted (via retract or toggle), the speed of the motor should be set to 0.

### PR DESCRIPTION
**Pull request by kchenturtles**
Created Intake Subsystem with methods for setting intake motor speed, toggling intake, retracting intake, putting intake in neutral, and extending intake

Closes #21

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    intakeMotor = new CANSparkMax(intakeMotorID, MotorType.kBrushless);
+    intakeMotor.setIdleMode(IdleMode.kBrake);
+  }
+
+  /**
+   * Sets the intake motor speed.
+   * 
+   * @param intakeSpeed The speed of the intake motor
+   */
+  public void setIntakeSpeed(double intakeSpeed) {
```
Comment (aidan-mundy) [inline src/main/java/frc/robot/subsystems/Intake.java]: Change to setSpinSpeed

IntakeSpeed is a bit redundant

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    if (solenoid.get() == DoubleSolenoid.Value.kReverse && intakeSpeed != 0) {
```
Comment (aidan-mundy) [inline src/main/java/frc/robot/subsystems/Intake.java]: add `isExtended`, `isRetracted` and `isNeutralExtension` methods that return booleans

Then use `isRetracted` here

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+      throw new IllegalStateException();
+    }
+    intakeMotor.set(intakeSpeed);
+  }
+
+  /**
+   * Toggles the {@link DoubleSolenoid.Value} between Forward and Reverse. If in neutral toggles it
```
Comment (aidan-mundy) [inline src/main/java/frc/robot/subsystems/Intake.java]: Change terminology from "Forward" and "Reverse" to "Extended" and "Retracted"

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      default:
+        retract();
+        return;
+    }
+  }
+
+  /**
+   * Sets the intake DoubleSolenoid to Reverse.
+   */
+  public void retract() {
```
Comment (aidan-mundy) [inline src/main/java/frc/robot/subsystems/Intake.java]:
```suggestion
  public void setRetracted() {
```

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        return;
+      case kForward:
+      default:
+        retract();
+        return;
+    }
+  }
+
+  /**
+   * Sets the intake DoubleSolenoid to Reverse.
```
Comment (aidan-mundy) [inline src/main/java/frc/robot/subsystems/Intake.java]:
```suggestion
   * Retracts the intake and sets spin speed to 0.
```

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+   */
+  public void retract() {
+    setIntakeSpeed(0.0);
+    solenoid.set(DoubleSolenoid.Value.kReverse);
+  }
+
+  /**
+   * Sets the intake DoubleSolenoid to neutral.
+   */
+  public void neutral() {
```
Comment (aidan-mundy) [inline src/main/java/frc/robot/subsystems/Intake.java]:
```suggestion
  public void setNeutralExtension() {
```

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  /**
+   * Sets the intake DoubleSolenoid to Reverse.
+   */
+  public void retract() {
+    setIntakeSpeed(0.0);
+    solenoid.set(DoubleSolenoid.Value.kReverse);
+  }
+
+  /**
+   * Sets the intake DoubleSolenoid to neutral.
```
Comment (aidan-mundy) [inline src/main/java/frc/robot/subsystems/Intake.java]:
```suggestion
   * Sets the intake to Neutral extension.
```

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+   */
+  public void neutral() {
+    solenoid.set(DoubleSolenoid.Value.kOff);
+  }
+
+  /**
+   * Sets the intake DoubleSolenoid to Forward.
```
Comment (aidan-mundy) [inline src/main/java/frc/robot/subsystems/Intake.java]:
```suggestion
   * Extends the intake.
```

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+   */
+  public void extend() {
```
Comment (aidan-mundy) [inline src/main/java/frc/robot/subsystems/Intake.java]:
```suggestion
  public void setExtended() {
```

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+  /**
+   * Constructs an Intake with a {@link CANSparkMax} at the given CAN ID and a
+   * {@link DoubleSolenoid} with the given forward and reverse channels.
+   *
+   * @param intakeMotorID The CAN ID of the intake motor
+   * @param forwardChannel The forward channel
+   * @param reverseChannel The reverse channel
+   */
+
+  public Intake(int intakeMotorID, int forwardChannel, int reverseChannel) {
```
Comment (aidan-mundy) [inline src/main/java/frc/robot/subsystems/Intake.java:30]: Please add another constructor that allows for a configurable MotorType

**Review by aidan-mundy — CHANGES_REQUESTED** [review]
Pretty good. A few changes to make to get everything sorted for merging

**Review by chit-sudeep — DISMISSED** [review]
Review intake review code and javadocs.

Looks good!

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    intakeMotor = new CANSparkMax(intakeMotorID, motorType);
+    intakeMotor.setIdleMode(IdleMode.kBrake);
+  }
+
+  /**
+   * Sets the intake motor speed.
+   * 
+   * @param intakeSpeed The speed of the intake motor
+   */
+  public void setSpinSpeed(double intakeSpeed) {
```
Comment (aidan-mundy) [inline src/main/java/frc/robot/subsystems/Intake.java]: Change "intakeSpeed" to "speed" or "newSpeed"

The method declaration should include a "throws" like this
```java
public void setSpinSpeed(double newSpeed) throws IllegalStateException {
```

Also, the javadoc should include a properly formatted "@throws" declaration

**Review by aidan-mundy — CHANGES_REQUESTED** [review]
Thanks for the updated code! Just one last change that I didn't notice before...

**Review by chit-sudeep — DISMISSED** [review]
Looked over all code and javadocs and if Aidan's suggestions were fixed.

Looks good!

File: src/main/java/frc/robot/subsystems/Intake.java [hunk 12]
```diff
@@ -50,12 +50,13 @@ public Intake(int intakeMotorID, MotorType motorType, int forwardChannel, int re
    * Sets the intake motor speed.
    * 
    * @param intakeSpeed The speed of the intake motor
```
Comment (aidan-mundy) [inline src/main/java/frc/robot/subsystems/Intake.java]: Changes to input variables should be reflected in the javadocs
```suggestion
   * @param newSpeed The speed of the intake motor
```

**Review by aidan-mundy — CHANGES_REQUESTED** [review]

**Review by chit-sudeep — APPROVED** [review]
Looked over all of the code, especially the Javadocs.

Looks good!

**Review by aidan-mundy — APPROVED** [review]

**Pull request merged**